bindings/rust/s2n-tls/src/raw/connection.rs

```diff
@@ -27,6 +30,21 @@ macro_rules! static_const_str {
     };
 }
 
+/// A trait indicating that a structure can function as a connection.
+pub trait AsConnection {
```

Comment:
This is basically a specialized version of [AsRef<Connection>](https://doc.rust-lang.org/std/convert/trait.AsRef.html) and [AsMut<Connection>](https://doc.rust-lang.org/std/convert/trait.AsMut.html). Is there a reason not to use those instead?

Reply:
I mostly hated writing AsRef<Connection> + AsMut<Connection> everywhere, since they always appear together 😅 `as_ref()` and `as_mut()` were also be tricky to call properly with Pins. But I switched back to AsRef/AsMut and it's not as messy as I remember.

Alternatively, I could keep AsConnection but write:
```
impl<T: AsRef<Connection> + AsMut<Connection>> AsConnection for T {
    fn conn(&self) -> &Connection {
        self.as_ref()
    }
    fn conn_mut(&mut self) -> &mut Connection {
        self.as_mut()
    }
}
```

But I am now leaning towards AsConnection being unnecessary noise. I just wish you could typedef type constraints :(

Reply:
You can actually make your own marker traits as well to cut down on the repetition:

```rust
pub trait AsConnection: AsRef<Connection> + AsMut<Connection> {}

impl<T: AsRef<Connection> + AsMut<Connection>> AsConnection for T {}
```

then you can just use `AsConnection` everywhere and it'll enforce the requirements:

```rust
fn foo<T: AsConnection>(t: &mut T) -> &mut Connection {
    t.as_mut()
}
```

Reply:
That would avoid writing AsRef+AsMut everywhere, but not handle the confusion with Pin's exisitng as_ref/as_mut. I think I'm going to leave it as is -- we can always add some marker traits to s2n-tokio later to avoid all the verbose type constraints.

Reply:
That's fine